### PR TITLE
W-11410177: Refactor: the current span should travel with the event

### PR DESCRIPTION
**Pull request by fsgonz**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventExecutionSpanProvider.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import org.mule.runtime.api.component.Component;
+import org.mule.runtime.core.api.config.MuleConfiguration;
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.core.internal.profiling.InternalSpan;
+
+/**
+ * Provides the Span from the core event and the component.
+ *
+ * @since 4.5.0
+ */
```
Comment (fsgonz) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventExecutionSpanProvider.java:19]: This should be a customizer of each attribute of the span (eventually increment it), not the creation of the span.

File: core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java [hunk 2]
```diff
@@ -52,7 +53,7 @@ public class DefaultProfilingService extends AbstractProfilingService {
   @Inject
   private ProfilingFeatureFlaggingService featureFlaggingService;
 
-  private SpanManager spanManager = new DefaultSpanManager();
+  private MuleCoreEventTracer eventTracer;
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java:56]: Rename to CoreEventTracer

File: core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java [hunk 3]
```diff
@@ -85,6 +86,13 @@ public void initialise() throws InitialisationException {
     super.initialise();
   }
 
+  private void createMuleEventTracer() {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java:89]: rename to createCoreEventTracer

File: core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java [hunk 4]
```diff
@@ -85,6 +86,13 @@ public void initialise() throws InitialisationException {
...
+    this.eventTracer = getMuleEventTracerBuilder()
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java:90]: Rename to getCoreEventTracerBuilder

File: core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java [hunk 5]
```diff
@@ -85,6 +86,13 @@ public void initialise() throws InitialisationException {
...
+        .withDefaultCoreEventExecutionSpanProvider(getMuleDefaultCoreEventExecutionSpanProvider())
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java:91]: Rename to getDefaultCoreEventExecutionSpanProvider

File: core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java [hunk 6]
```diff
@@ -168,8 +176,11 @@ public <S> Flux<S> setCurrentExecutionContext(Flux<S> original, Function<S, Exec
   }
 
   @Override
-  public SpanManager getSpanManager() {
-    return spanManager;
+  public MuleCoreEventTracer getMuleCoreEventTracer() {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingService.java:179]: rename to getCoreEventTracer

File: core/src/main/java/org/mule/runtime/core/internal/profiling/NoOpProfilingService.java [hunk 7]
```diff
@@ -128,7 +134,62 @@ public <S> Flux<S> setCurrentExecutionContext(Flux<S> original, Function<S, Exec
...
-  public SpanManager getSpanManager() {
-    return spanManager;
+  public MuleCoreEventTracer getMuleCoreEventTracer() {
+    return muleCoreEventTracer;
+  }
+
+  private static class NoOpCoreEventTracer implements MuleCoreEventTracer {
+
+    @Override
+    public InternalSpan startComponentExecutionSpan(CoreEvent coreEvent, Component component) {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/NoOpProfilingService.java:144]: Rename to startComponentSpan

File: core/src/main/java/org/mule/runtime/core/internal/profiling/NoOpProfilingService.java [hunk 8]
```diff
@@ -128,7 +134,62 @@ public <S> Flux<S> setCurrentExecutionContext(Flux<S> original, Function<S, Exec
...
+      return NoOpInternalSpan.INSTANCE;
+    }
+
+    @Override
+    public InternalSpan startComponentExecutionSpan(CoreEvent coreEvent, Component component,
+                                                    CoreEventSpanCustomizer coreEventSpanCustomizer) {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/NoOpProfilingService.java:150]: Rename to startComponentSpan

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/MuleCoreEventTracer.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * A {@link CoreEvent} is dispatched to a flow and begins to perform a Http Request. The flow is being processed and also the
+ * http:request, but the current context span is the span corresponding to the http request:
+ *
+ * ------------- Flow Span -----------------------------------------------------------> |________________ Http Request Span
+ * ------- (Current Context Span) ------------>
+ *
+ * @since 4.5.0
+ */
+public interface MuleCoreEventTracer {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/MuleCoreEventTracer.java:35]: Rename to CoreEventTracer

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventSpanUtils.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  private static final String CORE_EVENT_SPAN_NAME_SEPARATOR = ":";
+  public static final String UNKNOWN = "unknown";
+
+  /**
+   * @param eventContext the {@link EventContext} to extract the parent span from.
+   * @return the current context span
+   *
+   * @see MuleCoreEventTracer
+   */
+  public static Optional<InternalSpan> getCurrentContextSpan(EventContext eventContext) {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventSpanUtils.java:37]: Rename to getCurrentSpan

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventSpanUtils.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    return empty();
+  }
+
+  /**
+   * @param componentIdentifier the {@link ComponentIdentifier}
+   * @return the name for the {@link org.mule.runtime.api.profiling.tracing.Span}
+   */
+  public static String getCoreEventSpanName(ComponentIdentifier componentIdentifier) {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventSpanUtils.java:49]: Rename to getSpanName

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventExecutionSpanProvider.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+public interface CoreEventExecutionSpanProvider {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventExecutionSpanProvider.java:20]: Rename to CoreEventSpanProvider or CoreEventSpanFactory

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultMuleCoreEventTracer.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+import org.mule.runtime.core.internal.profiling.InternalSpan;
+import org.mule.runtime.core.internal.profiling.tracing.event.span.CoreEventSpanCustomizer;
+import org.mule.runtime.core.internal.profiling.tracing.event.tracer.MuleCoreEventTracer;
+
+/**
+ * A default implementation for a {@link MuleCoreEventTracer}.
+ *
+ * @since 4.5.0
+ */
+public class DefaultMuleCoreEventTracer implements MuleCoreEventTracer {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultMuleCoreEventTracer.java:24]: Rename to DefaultCoreEventTracer

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultMuleCoreEventTracer.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  public InternalSpan startComponentExecutionSpan(CoreEvent coreEvent, Component component) {
+    return setCurrentContextSpanInEventContextIfPossible(coreEvent,
+                                                         defaultCoreEventExecutionSpanProvider.getSpan(coreEvent, component,
+                                                                                                       muleConfiguration));
+  }
+
+  @Override
+  public InternalSpan startComponentExecutionSpan(CoreEvent coreEvent, Component component,
+                                                  CoreEventSpanCustomizer coreEventSpanCustomizer) {
+    return setCurrentContextSpanInEventContextIfPossible(coreEvent,
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultMuleCoreEventTracer.java:52]: Rename to setCurrentSpanIfPossible

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultMuleCoreEventTracer.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    if (eventContext instanceof DistributedTraceContextAware) {
+      ((DistributedTraceContextAware) eventContext)
+          .getDistributedTraceContext()
+          .setContextCurrentSpan(currentSpan);
+    }
+
+    return currentSpan;
+  }
+
+  private void endCurrentContextSpanIfPosibble(CoreEvent coreEvent) {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultMuleCoreEventTracer.java:75]: Rename to endCurrentSpanIfPossible

**Review by IvanAndresFritzler — CHANGES_REQUESTED** [review]

File: core-tests/src/test/java/org/mule/runtime/core/internal/profiling/consumer/tracing/DefaultMuleCoreEventTracerTestCase.java [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+import java.util.Optional;
+
+import io.qameta.allure.Feature;
+import io.qameta.allure.Story;
+import org.jetbrains.annotations.NotNull;
+import org.junit.Test;
+
+
+@Feature(PROFILING)
+@Story(DEFAULT_MULE_CORE_EVENT_TRACER)
```
Comment (IvanAndresFritzler) [inline core-tests/src/test/java/org/mule/runtime/core/internal/profiling/consumer/tracing/DefaultMuleCoreEventTracerTestCase.java:63]: Rename to DEFAULT_CORE_EVENT_TRACER

File: core-tests/src/test/java/org/mule/runtime/core/internal/profiling/consumer/tracing/DefaultMuleCoreEventTracerTestCase.java [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+public class DefaultMuleCoreEventTracerTestCase {
```
Comment (IvanAndresFritzler) [inline core-tests/src/test/java/org/mule/runtime/core/internal/profiling/consumer/tracing/DefaultMuleCoreEventTracerTestCase.java:64]: Rename to DefaultCoreEventTracerTestCase

File: core/src/main/java/org/mule/runtime/core/internal/execution/FlowProcessMediator.java [hunk 18]
```diff
@@ -143,6 +148,7 @@ public class FlowProcessMediator implements Initialisable {
   private NotificationHelper notificationHelper;
   private final List<SourceInterceptor> sourceInterceptors = new LinkedList<>();
   private Optional<CorrelationIdGenerator> correlationIdGenerator;
+  private CoreEventTracer muleEventTracer;
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/execution/FlowProcessMediator.java:151]: Rename to eventTracer

File: core/src/main/java/org/mule/runtime/core/internal/profiling/NoOpProfilingService.java [hunk 19]
```diff
@@ -31,7 +36,7 @@
  */
 public class NoOpProfilingService implements InternalProfilingService, PrivilegedProfilingService {
 
-  private SpanManager spanManager = new DefaultSpanManager();
+  private CoreEventTracer muleCoreEventTracer = new NoOpCoreEventTracer();
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/NoOpProfilingService.java:39]: Rename to eventTracer

File: core/src/main/java/org/mule/runtime/core/internal/profiling/OpentelemetryExecutionSpan.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapGetter;
+
+/**
+ * A Wrapper for a span that manages open telemetry spans.
+ *
+ * @since 4.5.0
+ */
+public class OpentelemetryExecutionSpan implements InternalSpan {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/OpentelemetryExecutionSpan.java:34]: Rename to OpenTelemetrySpan

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventSpanUtils.java [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+      return "unknown";
+    }
+
+    return componentIdentifier.getNamespace();
+  }
+
+  /**
+   * @return the default {@link CoreEventSpanFactory}
+   */
+  public static CoreEventSpanFactory getMuleDefaultCoreEventExecutionSpanProvider() {
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/span/CoreEventSpanUtils.java:73]: Rename to getDefaultCoreEventSpanFactory

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+import org.mule.runtime.core.internal.profiling.tracing.event.tracer.CoreEventTracer;
+
+/**
+ * A default implementation for a {@link CoreEventTracer}.
+ *
+ * @since 4.5.0
+ */
+public class DefaultCoreEventTracer implements CoreEventTracer {
+
+  private final CoreEventSpanFactory defaultCoreEventExecutionSpanProvider;
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java:26]: Rename to defaultCoreEventSpanFactory

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+
+  /**
+   * A Builder for a {@link DefaultEventTracerBuilder}.
+   *
+   * @since 4.5.0
+   */
+  public static final class DefaultEventTracerBuilder {
+
+    private MuleConfiguration muleConfiguration;
+    private CoreEventSpanFactory coreEventExecutionSpanProvider;
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java:92]: Rename to coreEventSpanFactory

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+
+    public DefaultEventTracerBuilder withMuleConfiguration(MuleConfiguration muleConfiguration) {
+      this.muleConfiguration = muleConfiguration;
+      return this;
+    }
+
+    public DefaultEventTracerBuilder withDefaultCoreEventExecutionSpanProvider(
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java:99]: rename to withDefaultCoreEventSpanFactory

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+                                         CoreEventSpanCustomizer coreEventSpanCustomizer) {
+    return endCurrentSpanIfPossible(coreEvent,
+                                    defaultCoreEventExecutionSpanProvider.getSpan(coreEvent, component,
+                                                                                  muleConfiguration,
+                                                                                  coreEventSpanCustomizer));
+  }
+
+  @Override
+  public void endCurrentSpan(CoreEvent coreEvent) {
+    endCurrentContextSpanIfPosibble(coreEvent);
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java:60]: rename to endCurrentSpanIfPossible

File: core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+
+    return currentSpan;
+  }
+
+  private void endCurrentContextSpanIfPosibble(CoreEvent coreEvent) {
+    EventContext eventContext = coreEvent.getContext();
+    if (eventContext instanceof DistributedTraceContextAware) {
+      ((DistributedTraceContextAware) eventContext)
+          .getDistributedTraceContext()
+          .endCurrentContextSpan();
```
Comment (IvanAndresFritzler) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/tracing/event/tracer/impl/DefaultCoreEventTracer.java:80]: rename to getCurrentSpan

**Review by IvanAndresFritzler — CHANGES_REQUESTED** [review]

**Review by IvanAndresFritzler — APPROVED** [review]

**Pull request merged**